### ISSUE TITLE
Support Mixture of Expert (MoE) Models

### ISSUE THREAD

**Issue by AlexCheema** [exo-explore/exo#32]
![IMG_0084](https://github.com/user-attachments/assets/b52fd7a3-1b0d-48b5-9f61-afc5931e1d18)

**Comment by AlexCheema** [comment]
![IMG_0085](https://github.com/user-attachments/assets/d55d16ac-476d-4f13-bff2-aba23d4bd035)

**Comment by mzbac** [comment]
I looked at this yesterday, it should be very similar to the llama sharding in the DeepseekV2DecoderLayer. But maybe worth trying model parallelism -> https://github.com/ml-explore/mlx-examples/pull/890

**Comment by 345ishaan** [comment]
will like to work on this :)

**Comment by AlexCheema** [comment]
> will like to work on this :)

@345ishaan that would be great - go for it

**Comment by mintisan** [comment]
Indeed, MoE is the most suitable application scenario for exo and should be prioritized for implementation.
Really looking forward to it

**Comment by youmego** [comment]
looking forward to MoE deepseek v2 total:236B active:21B
<html>
<body>

Model | #Total Params | #Activated Params | Context Length | Download
-- | -- | -- | -- | --
DeepSeek-V2 | 236B | 21B | 128k | 🤗 HuggingFace
DeepSeek-V2-Chat (RL) | 236B | 21B | 128k | 🤗 HuggingFace


</body>
</html>Model	#Total Params	#Activated Params	Context Length	Download
DeepSeek-V2-Lite	16B	2.4B	32k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite)
DeepSeek-V2-Lite-Chat (SFT)	16B	2.4B	32k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite-Chat)
DeepSeek-V2	236B	21B	128k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2)
DeepSeek-V2-Chat (RL)	236B	21B	128k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Chat)

**Comment by youmego** [comment]
+-------------------+------------+-------------------+--------------+----------------+
| Model             | #Total Params | #Activated Params | Context Length | Download       |
+-------------------+------------+-------------------+--------------+----------------+
| DeepSeek-V2       | 236B         | 21B               | 128k          | 🤗 HuggingFace |
| DeepSeek-V2-Chat (RL) | 236B     | 21B               | 128k          | 🤗 HuggingFace |
+-------------------+------------+-------------------+--------------+----------------+

**Comment by youmego** [comment]
`+-------------------+------------+-------------------+--------------+----------------+
| Model             | #Total Params | #Activated Params | Context Length | Download       |
+-------------------+------------+-------------------+--------------+----------------+
| DeepSeek-V2-Lite  | 16B          | 2.4B              | 32k           | 🤗 HuggingFace |
| DeepSeek-V2-Lite-Chat (SFT) | 16B | 2.4B              | 32k           | 🤗 HuggingFace |
| DeepSeek-V2       | 236B         | 21B               | 128k          | 🤗 HuggingFace |
| DeepSeek-V2-Chat (RL) | 236B     | 21B               | 128k          | 🤗 HuggingFace |
+-------------------+------------+-------------------+--------------+----------------+`

**Comment by youmego** [comment]
`<html>
<body>

Model | #Total Params | #Activated Params | Context Length | Download
-- | -- | -- | -- | --
DeepSeek-V2-Lite | 16B | 2.4B | 32k | 🤗 HuggingFace
DeepSeek-V2-Lite-Chat (SFT) | 16B | 2.4B | 32k | 🤗 HuggingFace
DeepSeek-V2 | 236B | 21B | 128k | 🤗 HuggingFace
DeepSeek-V2-Chat (RL) | 236B | 21B | 128k | 🤗 HuggingFace


</body>
</html>Model	#Total Params	#Activated Params	Context Length	Download
DeepSeek-V2-Lite	16B	2.4B	32k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite)
DeepSeek-V2-Lite-Chat (SFT)	16B	2.4B	32k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Lite-Chat)
DeepSeek-V2	236B	21B	128k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2)
DeepSeek-V2-Chat (RL)	236B	21B	128k	[🤗 HuggingFace](https://huggingface.co/deepseek-ai/DeepSeek-V2-Chat)`

**Comment by 345ishaan** [comment]
> looking forward to support MoE deepseek v2 total:236B active:21B +--------------------------+---------------+---------------------+-----------------+-------------------+ | Model | #Total Params | #Activated Params | Context Length | Download | +--------------------------+---------------+---------------------+-----------------+-------------------+ | DeepSeek-V2 | 236B | 21B | 128k | 🤗 HuggingFace | | DeepSeek-V2-Chat (RL) | 236B | 21B | 128k | 🤗 HuggingFace | +--------------------------+---------------+------------------ -+-----------------+------------------+

yeah i was planning to experiment the setup with  https://github.com/deepseek-ai/DeepSeek-Coder-V2 . Will be looking into it this weekend.

**Comment by ChaseKolozsy** [comment]
@AlexCheema 

Is it possible to have the active parameters favor an Nvidia Cuda GPU and let the other nodes store the inactive parameters?

**Comment by AlexCheema** [comment]
> @AlexCheema
> 
> Is it possible to have the active parameters favor an Nvidia Cuda GPU and let the other nodes store the inactive parameters?

Yes! We can definitely do something like this.
We are moving towards more of a general distributed AI framework that will enable things like this.

**Comment by ChaseKolozsy** [comment]
> > @AlexCheema
> > Is it possible to have the active parameters favor an Nvidia Cuda GPU and let the other nodes store the inactive parameters?
> 
> Yes! We can definitely do something like this. We are moving towards more of a general distributed AI framework that will enable things like this.

That's awesome!

Your project has gotten really excited, especially after I saw you get deepseek v3 running on 8 mac minis and it even inspired me to splurge and purchase a subscription to openai's pro subscription because I thought I was a genius and could somehow get one computer to listen to three data streams from three thunderbolt connections in parallel and triple the throughput, and I though o1 pro could help me do that, but it gave me a quick education that this is not practical as the ports share the same controller and the controller itself is limited by its bandwidth capacity. But I have been very curious and asking it lots of questions and one thing it kept mentioning is the possibility of replicating the hottest or most utilized experts.

**_Will exo be able to manage expert replication such that if an expert is the cause of the bottle neck due to over utilization, will exo eventually be able to detect this and replicate that node to ease the bottleneck?_**

**Comment by ChaseKolozsy** [comment]
> > @AlexCheema
> > Is it possible to have the active parameters favor an Nvidia Cuda GPU and let the other nodes store the inactive parameters?
> 
> Yes! We can definitely do something like this. We are moving towards more of a general distributed AI framework that will enable things like this.

https://blog.exolabs.net/day-9/

Nice!

**Comment by rltakashige** [comment]
We do support MoEs in EXO 1.0, like Kimi K2 Thinking.
